Export datasets to `Float32` by default

mtanneau opened the pull request:
closes #162 
There are multiple ways of doing this, I opted to implement it at the level of the HDF5 writer.
Pros:
* requires minimal changes to the rest of the source code (especially code to extract primal/dual solutions)
* ensures that _all_ floating point data we export is in the prescribed arithmetic
* backward-compatible
* allows 

Cons:
* only happens when we export to HDF5
* forces _all_ floating-point data to be exported in the same arithmetic. If users have parts of the dataset that they want to keep in, e.g., `Float64`... they can't.

Alternative approaches:
* move the part of the code that handles the conversion to `Float32` to a stand-alone function, similar to how [ML4OPF does it](https://github.com/AI4OPT/ML4OPF/blob/499a982a8d2e55c8964fb08441094461ed36ab3b/ml4opf/parsers/pglearn.py#L176). This would allow to convert data outside of the HDF5 pipeline.
  [EDIT: I implemented this functionality instead]
* Only convert primal/dual solutions to `Float32`. I think this would make a lot of code very cumbersome and repetitive, and it would miss anything that's not primal/dual solution (e.g. input data)

TODO:
* [x] validate proposed approach
* [x] add/expand unit tests
* [x] update sampler script & slurm template
* [x] update config workflow to let users specify a floating-point arithmetic

mtanneau commented:
I kinda like the first alternative approach (stand-alone function to convert all floating point data, outside of HDF5).
However, AFAICT, this would need to create a new dictionary, because otherwise Julia's type system may throw errors or (worse) make it silently fail.

For instance, consider the simple code
```julia
julia> function fp64_to_fp32(D)
    for (k, v) in D
        if typeof(v) == Float64
            D[k] = Float32(v)
        end
    end
    return D
end

julia> d = Dict("a" => 1.0, "b" => 2.0)
Dict{String, Float64} with 2 entries:
  "b" => 2.0
  "a" => 1.0
julia> fp64_to_fp32(d)
Dict{String, Float64} with 2 entries:
  "b" => 2.0
  "a" => 1.0
```
^^ because `d` is a `Dict{String,Float64}`, Julia silently re-converts everything back to `Float64` when executing `D[k] = Float32(v)` inside the call to `fp64_to_fp32`.

klamike commented:
Yeah that is annoying. I think what you have now is fine though. If somebody really needs a few keys to be Float64, they can dev OPFGenerator and add an escape hatch in `save_h5` e.g. `if k == "pg": <skip conversion>`

mtanneau commented:
I reverted my previous changes, and added a separate function (see `src/utils/float.jl`) to convert floating-point data in a dictionary. Note that I only convert scalar and dense array data.

mtanneau commented:
I run some additional local tests to verify that the overall pipeline functions as intended.
In particular, I want to make sure that nothing gets accidentally re-converted to `Float64` when we execute post-processing.

mtanneau commented on exp/sampler.jl:
```diff
@@ -22,6 +22,8 @@ using MathOptSymbolicAD
 
 using OPFGenerator
 
+const DEFAULT_FLOAT_PRECISION = Float32
```
This is where we set the floating-point precision to be 32-bit by default...

mtanneau commented on exp/sampler.jl:
```diff
@@ -90,6 +93,17 @@ if abspath(PROGRAM_FILE) == @__FILE__
     smax = parse(Int, ARGS[3])
 
+    # Grab precision used for exporting data
+    fp_string = get(config, "floating_point_type", "$(DEFAULT_FLOAT_PRECISION)")
+    float_type = if lowercase(fp_string) == "float32"
+        Float32
+    elseif lowercase(fp_string) == "float64"
+        Float64
+    else
+        error("Invalid floating-point type: $(fp_string); only `Float32` and `Float64` are supported.")
+    end
+    @info "Floating-point data will be exported in `$(float_type)`"
```
... combined with this code where we extract floating-point type...

mtanneau commented:
I confirm that the current code results in post-processed datasets where floating-point data is in `Float32`. 

As a nice side-effect, we get a little better than 2x reduction in archived dataset size (at least for the 14-bus system).
Last check to be performed locally: verify that the float32 data matches well the data that was exported in float64. There's no reason why it shouldn't... but we never know

mtanneau commented:
@klamike ready for your review.
Please let me know if you'd like some changes to how the config files are handled.

klamike reviewed (approved):
LGTM
